File: snippets/algorithms/graph-traversal/graph_traversal.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <chrono>
#include <thread>

/**
 * Graph Traversal Algorithms in C++
 * 
 * This program demonstrates BFS and DFS traversal algorithms on a graph.
 * 
 * Compile: g++ -std=c++11 graph_traversal.cpp -o graph_traversal
 * Run: ./graph_traversal
 */

class Graph {
private:
    // Adjacency list representation
    std::unordered_map<std::string, std::vector<std::string>> adjacencyList;
    
    // Helper function to get sorted neighbors for consistent output
    std::vector<std::string> getSortedNeighbors(const std::string& vertex) const {
        std::vector<std::string> neighbors = adjacencyList.at(vertex);
        std::sort(neighbors.begin(), neighbors.end());
        return neighbors;
    }
    
    // Helper for recursive DFS
    void dfsHelper(const std::string& vertex, std::unordered_set<std::string>& visited, 
                   std::vector<std::string>& result) const {
        // Mark as visited and add to result
        visited.insert(vertex);
        result.push_back(vertex);
        
        std::cout << "Visiting: " << vertex << std::endl;
        std::cout << "Visited so far: [";
        for (size_t i = 0; i < result.size(); ++i) {
            std::cout << result[i];
            if (i < result.size() - 1) std::cout << ", ";
        }
        std::cout << "]" << std::endl;
        std::cout << "------------------------------" << std::endl;
        
        // Pause for demonstration
        std::this_thread::sleep_for(std::chrono::milliseconds(500));
        
        // Get sorted neighbors for consistent order
        std::vector<std::string> neighbors = getSortedNeighbors(vertex);
        
        // Recursively visit all unvisited neighbors
        for (const auto& neighbor : neighbors) {
            if (visited.find(neighbor) == visited.end()) {
                dfsHelper(neighbor, visited, result);
            }
        }
    }
    
public:
    // Add a vertex to the graph
    void addVertex(const std::string& vertex) {
        if (adjacencyList.find(vertex) == adjacencyList.end()) {
            adjacencyList[vertex] = std::vector<std::string>();
        }
    }
    
    // Add an edge between two vertices
    void addEdge(const std::string& v1, const std::string& v2) {
        // Ensure both vertices exist
        addVertex(v1);
        addVertex(v2);
        
        // Add the edge (undirected graph)
        adjacencyList[v1].push_back(v2);
        adjacencyList[v2].push_back(v1);
    }
    
    // Breadth-First Search traversal
    std::vector<std::string> bfs(const std::string& start) const {
        if (adjacencyList.find(start) == adjacencyList.end()) {
            return {};
        }
        
        std::unordered_set<std::string> visited;
        std::queue<std::string> queue;
        std::vector<std::string> result;
        
        // Initialize with starting vertex
        visited.insert(start);
        queue.push(start);
        
        std::cout << "Starting BFS traversal from vertex " << start << std::endl;
        
        while (!queue.empty()) {
            // Dequeue the first vertex
            std::string vertex = queue.front();
            queue.pop();
            result.push_back(vertex);
            
            std::cout << "Visiting: " << vertex << std::endl;
            
            // Print queue contents
            std::cout << "Queue: [";
            std::queue<std::string> queueCopy = queue;
            bool first = true;
            while (!queueCopy.empty()) {
                if (!first) std::cout << ", ";
                std::cout << queueCopy.front();
                queueCopy.pop();
                first = false;
            }
            std::cout << "]" << std::endl;
            
            // Print visited vertices
            std::cout << "Visited so far: [";
            for (size_t i = 0; i < result.size(); ++i) {
                std::cout << result[i];
                if (i < result.size() - 1) std::cout << ", ";
            }
            std::cout << "]" << std::endl;
            std::cout << "------------------------------" << std::endl;
            
            // Pause for demonstration
            std::this_thread::sleep_for(std::chrono::milliseconds(500));
            
            // Get sorted neighbors for consistent order
            std::vector<std::string> neighbors = getSortedNeighbors(vertex);
            
            // Enqueue all unvisited neighbors
            for (const auto& neighbor : neighbors) {
                if (visited.find(neighbor) == visited.end()) {
                    visited.insert(neighbor);
                    queue.push(neighbor);
                }
            }
        }
        
        return result;
    }
    
    // Depth-First Search traversal (recursive)
    std::vector<std::string> dfsRecursive(const std::string& start) const {
        if (adjacencyList.find(start) == adjacencyList.end()) {
            return {};
        }
        
        std::unordered_set<std::string> visited;
        std::vector<std::string> result;
        
        std::cout << "Starting recursive DFS traversal from vertex " << start << std::endl;
        
        dfsHelper(start, visited, result);
        
        return result;
    }
    
    // Depth-First Search traversal (iterative)
    std::vector<std::string> dfsIterative(const std::string& start) const {
        if (adjacencyList.find(start) == adjacencyList.end()) {
            return {};
        }
        
        std::unordered_set<std::string> visited;
        std::stack<std::string> stack;
        std::vector<std::string> result;
        
        // Initialize with starting vertex
        stack.push(start);
        
        std::cout << "Starting iterative DFS traversal from vertex " << start << std::endl;
        
        while (!stack.empty()) {
            // Pop the top vertex
            std::string vertex = stack.top();
            stack.pop();
            
            // If not visited, process it
            if (visited.find(vertex) == visited.end()) {
                visited.insert(vertex);
                result.push_back(vertex);
                
                std::cout << "Visiting: " << vertex << std::endl;
                
                // Print stack contents (in reverse order since it's LIFO)
                std::cout << "Stack: [";
                std::vector<std::string> stackContents;
                std::stack<std::string> stackCopy = stack;
                while (!stackCopy.empty()) {
                    stackContents.push_back(stackCopy.top());
                    stackCopy.pop();
                }
                std::reverse(stackContents.begin(), stackContents.end());
                for (size_t i = 0; i < stackContents.size(); ++i) {
                    std::cout << stackContents[i];
                    if (i < stackContents.size() - 1) std::cout << ", ";
                }
                std::cout << "]" << std::endl;
                
                // Print visited vertices
                std::cout << "Visited so far: [";
                for (size_t i = 0; i < result.size(); ++i) {
                    std::cout << result[i];
                    if (i < result.size() - 1) std::cout << ", ";
                }
                std::cout << "]" << std::endl;
                std::cout << "------------------------------" << std::endl;
                
                // Pause for demonstration
                std::this_thread::sleep_for(std::chrono::milliseconds(500));
                
                // Get sorted neighbors in reverse order for stack
                std::vector<std::string> neighbors = getSortedNeighbors(vertex);
                std::reverse(neighbors.begin(), neighbors.end());
                
                // Push all unvisited neighbors onto the stack
                for (const auto& neighbor : neighbors) {
                    if (visited.find(neighbor) == visited.end()) {
                        stack.push(neighbor);
                    }
                }
            }
        }
        
        return result;
    }
    
    // Print the graph structure
    void visualizeGraph() const {
        std::cout << "\nGraph Structure:" << std::endl;
        std::cout << "------------------------------" << std::endl;
        
        // Sort vertices for consistent output
        std::vector<std::string> vertices;
        for (const auto& pair : adjacencyList) {
            vertices.push_back(pair.first);
        }
        std::sort(vertices.begin(), vertices.end());
        
        for (const auto& vertex : vertices) {
            std::vector<std::string> neighbors = getSortedNeighbors(vertex);
            
            std::cout << vertex << " -> [";
            for (size_t i = 0; i < neighbors.size(); ++i) {
                std::cout << neighbors[i];
                if (i < neighbors.size() - 1) std::cout << ", ";
            }
            std::cout << "]" << std::endl;
        }
        
        std::cout << "------------------------------" << std::endl;
    }
};

// Create a sample graph for demonstration
Graph createSampleGraph() {
    Graph g;
    
    // Add edges to build this graph:
    //     A
    //    / \
    //   B   C
    //  / \   \
    // D   E---F
    
    std::vector<std::pair<std::string, std::string>> edges = {
        {"A", "B"}, {"A", "C"},
        {"B", "D"}, {"B", "E"},
        {"C", "F"}, {"E", "F"}
    };
    
    for (const auto& edge : edges) {
        g.addEdge(edge.first, edge.second);
    }
    
    return g;
}

// Main function
int main() {
    // Create a sample graph
    Graph g = createSampleGraph();
    g.visualizeGraph();
    
    // Demonstrate BFS
    std::cout << "\n=== BFS Traversal ===" << std::endl;
    std::vector<std::string> bfsResult = g.bfs("A");
    std::cout << "BFS Result: [";
    for (size_t i = 0; i < bfsResult.size(); ++i) {
        std::cout << bfsResult[i];
        if (i < bfsResult.size() - 1) std::cout << ", ";
    }
    std::cout << "]" << std::endl;
    
    // Demonstrate recursive DFS
    std::cout << "\n=== DFS Traversal (Recursive) ===" << std::endl;
    std::vector<std::string> dfsRecResult = g.dfsRecursive("A");
    std::cout << "DFS Recursive Result: [";
    for (size_t i = 0; i < dfsRecResult.size(); ++i) {
        std::cout << dfsRecResult[i];
        if (i < dfsRecResult.size() - 1) std::cout << ", ";
    }
    std::cout << "]" << std::endl;
    
    // Demonstrate iterative DFS
    std::cout << "\n=== DFS Traversal (Iterative) ===" << std::endl;
    std::vector<std::string> dfsIterResult = g.dfsIterative("A");
    std::cout << "DFS Iterative Result: [";
    for (size_t i = 0; i < dfsIterResult.size(); ++i) {
        std::cout << dfsIterResult[i];
        if (i < dfsIterResult.size() - 1) std::cout << ", ";
    }
    std::cout << "]" << std::endl;
    
    return 0;
} 

File: snippets/algorithms/graph-traversal/GraphTraversal.java
import java.util.*;

/**
 * Graph Traversal Algorithms in Java
 * 
 * This class demonstrates BFS and DFS traversal algorithms on a graph.
 * 
 * Compile: javac GraphTraversal.java
 * Run: java GraphTraversal
 */
public class GraphTraversal {
    
    /**
     * Graph class representing a graph using adjacency list
     */
    static class Graph {
        // Adjacency list representation
        private Map<String, List<String>> adjacencyList;
        
        /**
         * Initialize an empty graph
         */
        public Graph() {
            this.adjacencyList = new HashMap<>();
        }
        
        /**
         * Add a vertex to the graph
         * @param vertex The vertex to add
         */
        public void addVertex(String vertex) {
            if (!adjacencyList.containsKey(vertex)) {
                adjacencyList.put(vertex, new ArrayList<>());
            }
        }
        
        /**
         * Add an edge between two vertices
         * @param v1 First vertex
         * @param v2 Second vertex
         */
        public void addEdge(String v1, String v2) {
            // Ensure both vertices exist
            addVertex(v1);
            addVertex(v2);
            
            // Add the edge (undirected graph)
            adjacencyList.get(v1).add(v2);
            adjacencyList.get(v2).add(v1);
        }
        
        /**
         * Helper method to get sorted neighbors for consistent output
         * @param vertex The vertex to get neighbors for
         * @return Sorted list of neighbors
         */
        private List<String> getSortedNeighbors(String vertex) {
            List<String> neighbors = new ArrayList<>(adjacencyList.get(vertex));
            Collections.sort(neighbors);
            return neighbors;
        }
        
        /**
         * Breadth-First Search traversal
         * @param start Starting vertex
         * @return List of vertices in the order they were visited
         */
        public List<String> bfs(String start) {
            if (!adjacencyList.containsKey(start)) {
                return new ArrayList<>();
            }
            
            Set<String> visited = new HashSet<>();
            Queue<String> queue = new LinkedList<>();
            List<String> result = new ArrayList<>();
            
            // Initialize with starting vertex
            visited.add(start);
            queue.add(start);
            
            System.out.println("Starting BFS traversal from vertex " + start);
            
            while (!queue.isEmpty()) {
                // Dequeue the first vertex
                String vertex = queue.poll();
                result.add(vertex);
                
                System.out.println("Visiting: " + vertex);
                System.out.println("Queue: " + queue);
                System.out.println("Visited so far: " + result);
                System.out.println("------------------------------");
                
                // Pause for demonstration
                try {
                    Thread.sleep(500);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                
                // Get sorted neighbors for consistent order
                List<String> neighbors = getSortedNeighbors(vertex);
                
                // Enqueue all unvisited neighbors
                for (String neighbor : neighbors) {
                    if (!visited.contains(neighbor)) {
                        visited.add(neighbor);
                        queue.add(neighbor);
                    }
                }
            }
            
            return result;
        }
        
        /**
         * Depth-First Search traversal (recursive)
         * @param start Starting vertex
         * @return List of vertices in the order they were visited
         */
        public List<String> dfsRecursive(String start) {
            if (!adjacencyList.containsKey(start)) {
                return new ArrayList<>();
            }
            
            Set<String> visited = new HashSet<>();
            List<String> result = new ArrayList<>();
            
            System.out.println("Starting recursive DFS traversal from vertex " + start);
            
            dfsHelper(start, visited, result);
            
            return result;
        }
        
        /**
         * Helper method for recursive DFS
         * @param vertex Current vertex
         * @param visited Set of visited vertices
         * @param result List to store the traversal result
         */
        private void dfsHelper(String vertex, Set<String> visited, List<String> result) {
            // Mark as visited and add to result
            visited.add(vertex);
            result.add(vertex);
            
            System.out.println("Visiting: " + vertex);
            System.out.println("Visited so far: " + result);
            System.out.println("------------------------------");
            
            // Pause for demonstration
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            
            // Get sorted neighbors for consistent order
            List<String> neighbors = getSortedNeighbors(vertex);
            
            // Recursively visit all unvisited neighbors
            for (String neighbor : neighbors) {
                if (!visited.contains(neighbor)) {
                    dfsHelper(neighbor, visited, result);
                }
            }
        }
        
        /**
         * Depth-First Search traversal (iterative)
         * @param start Starting vertex
         * @return List of vertices in the order they were visited
         */
        public List<String> dfsIterative(String start) {
            if (!adjacencyList.containsKey(start)) {
                return new ArrayList<>();
            }
            
            Set<String> visited = new HashSet<>();
            Stack<String> stack = new Stack<>();
            List<String> result = new ArrayList<>();
            
            // Initialize with starting vertex
            stack.push(start);
            
            System.out.println("Starting iterative DFS traversal from vertex " + start);
            
            while (!stack.isEmpty()) {
                // Pop the top vertex
                String vertex = stack.pop();
                
                // If not visited, process it
                if (!visited.contains(vertex)) {
                    visited.add(vertex);
                    result.add(vertex);
                    
                    System.out.println("Visiting: " + vertex);
                    System.out.println("Stack: " + stack);
                    System.out.println("Visited so far: " + result);
                    System.out.println("------------------------------");
                    
                    // Pause for demonstration
                    try {
                        Thread.sleep(500);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                    
                    // Get sorted neighbors in reverse order for stack
                    List<String> neighbors = getSortedNeighbors(vertex);
                    Collections.reverse(neighbors);
                    
                    // Push all unvisited neighbors onto the stack
                    for (String neighbor : neighbors) {
                        if (!visited.contains(neighbor)) {
                            stack.push(neighbor);
                        }
                    }
                }
            }
            
            return result;
        }
        
        /**
         * Print the graph structure
         */
        public void visualizeGraph() {
            System.out.println("\nGraph Structure:");
            System.out.println("------------------------------");
            
            // Sort vertices for consistent output
            List<String> vertices = new ArrayList<>(adjacencyList.keySet());
            Collections.sort(vertices);
            
            for (String vertex : vertices) {
                List<String> neighbors = getSortedNeighbors(vertex);
                System.out.println(vertex + " -> " + neighbors);
            }
            
            System.out.println("------------------------------");
        }
    }
    
    /**
     * Create a sample graph for demonstration
     * @return A sample graph
     */
    public static Graph createSampleGraph() {
        Graph g = new Graph();
        
        // Add edges to build this graph:
        //     A
        //    / \
        //   B   C
        //  / \   \
        // D   E---F
        
        String[][] edges = {
            {"A", "B"}, {"A", "C"},
            {"B", "D"}, {"B", "E"},
            {"C", "F"}, {"E", "F"}
        };
        
        for (String[] edge : edges) {
            g.addEdge(edge[0], edge[1]);
        }
        
        return g;
    }
    
    /**
     * Main method to demonstrate graph traversal algorithms
     * @param args Command line arguments (not used)
     */
    public static void main(String[] args) {
        // Create a sample graph
        Graph g = createSampleGraph();
        g.visualizeGraph();
        
        // Demonstrate BFS
        System.out.println("\n=== BFS Traversal ===");
        List<String> bfsResult = g.bfs("A");
        System.out.println("BFS Result: " + bfsResult);
        
        // Demonstrate recursive DFS
        System.out.println("\n=== DFS Traversal (Recursive) ===");
        List<String> dfsRecResult = g.dfsRecursive("A");
        System.out.println("DFS Recursive Result: " + dfsRecResult);
        
        // Demonstrate iterative DFS
        System.out.println("\n=== DFS Traversal (Iterative) ===");
        List<String> dfsIterResult = g.dfsIterative("A");
        System.out.println("DFS Iterative Result: " + dfsIterResult);
    }
} 